### ISSUE TITLE
Analyze blunders

### ISSUE THREAD

**Issue by mooskagh** [glinscott/leela-chess#558]
# Important!

When reporting positions to analyze, please use the following form. It makes it easier to see what's problematic with the position:

- **Id:** Optional unique ID. Come up with something. :) number/word, just to make it easier to refer this position in further comments.
- **Game:**  Preferably link to lichess.org (use lichess.org/paste), or at least PGN text.
- **Bad move:**  Bad move with number, and optionally stockfish eval for that move
- **Correct move:** Good move to play, optionally with stockfish eval
- **Screenshot:** optional, screenshot of the position, pasted right into the message (not as link!). Helps grasping the problem without following links
- **Configuration**: Configuration used, including `lc0`/`lczero` version, operating system, and non-default parameters (number of threads, batch size, fpu reduction, etc).
- **Network ID**: Network ID, very important
- **Time control**:  Time control used, and if known how much time/nodes was spent thinking this move
- **Comments**: any comments that you may have, e.g. free word explanation what's happening in position.

# (old text below)

There are many reports on forums asking about blunders, and the answers so far had been something along the lines "it's fine, it will learn eventually, we don't know exactly why it happens".

I think at this point it makes sense to actually look into them to confirm that there no some blind spots in training. For that we need to:

- Open position in engine and check counters
- Try several times to evaluate that move with training configuration "800 playouts, with Dirichlet noise and temperature (`--temperature=1.0 --noise`)" to see how training data would look like for this position.

Eventually all of this would be nice to have as a single command, but we can start manually.

For `lc0`, that can be done this way: `--verbose-move-stats -t 1 --minibatch-size=1 --no-smart-pruning` (unless you want to debug specifically with other settings).

Then run UCI interface, do command:
```
position startpos moves e2e4 ....
```
(PGN move to UCI notation can be converted using `pgn-extract -Wuci`)

Then do:
```
go nodes 10
```
see results, add some more nodes by running:
```
go nodes 20
go nodes 100
go nodes 800
go nodes 5000
go nodes 10000
and so on
```
And look how counters change.

Counters:
```
e2e4 N: 329 (+ 4) (V: -12.34%) (P:38.12%) (Q: -0.2325) (U: 0.2394) (Q+U: 0.0069)
 ^      ^    ^      ^           ^          ^            ^           ^
 |      |    |      |           |          |            |           Q+U, see below
 |      |    |      |           |          |           U from PUCT formula,
 |      |    |      |           |          |           see below.
 |      |    |      |           |         Average value of V in a subtree
 |      |    |      |          Probability of this move, from NN, but if Dirichlet
 |      |    |      |          node is on, it's also added here, 0%..100%
 |      |    |     Expected outcome for this position, directly from NN, -100%..100%
 |      |   How many visits are processed by other threads when this is printed.
 |     Number of visits. The move with maximum visits is chosen for play.
Move

* U = P * Cpuct * sqrt(sum of N of all moves) / (N + 1)
  CPuct is a search parameter, can be changed with a command line flag.
* The move with largest Q+U will be visited next
```

# Help wanted:
- Feel free to post positions that you think need analyzing (don't forget to also mention network Id used, and also all other settings are nice to know)
- Feel free to analyze what other people posted

**Comment by Tilps** [comment]
https://lichess.org/e07JvP6g - I started analyzing this - the position after the blunder has a 0.11% policy for a move which is checkmate. Takes 20k visits to get its first look and then it obviously gets every visit.  I haven't tested how that varies with noise applied.

**Comment by Ghotrix** [comment]
isn't position startpos fen more convenient for this case?

**Comment by dubslow** [comment]
Here's an easy one ply discovered attack tactic missed by Leela after 2K5 nodes. Position: https://lichess.org/mbWjiT93#105 Twitch recording of the thinking time/engine output: https://clips.twitch.tv/GenerousSmellyEggnogPunchTrees And as the lichess analysis says, this was "merely" the cherry on top of the multiple-mistakes cake. How to swing 15 points eval in just 3 moves!

Further analysis requested please. How many playouts until Leela even once searches the tactic?

**Comment by hsntgm** [comment]
@mooskagh thanks for diagram.

If i wrong please correct me.

Leela's brain gets power from memorized games and positional samples she collected in self play and we call it **visits**.I see she has **visits** comes from weights instead of alfa-beta pruning.If there is a tactical opportunity in the position but  leela visits an other move much she choose it.

In basic tactical positions occurs suddenly in the game and the hardest part is to teach her this.
Is it necessary to play billions of games in order to learn the tactical motifs that occur during the game?

Or can you add a simple tactical search algorithm triggers on every move working independently from visits.

```
e2e4 N: 329 (+ 4) (V: -12.34%) (P:38.12%) (Q: -0.2325) (U: 0.2394) (Q+U: 0.0069)
 ^      ^    ^      ^           ^          ^            ^           ^
 |      |    |      |           |          |            |           Q+U, see below
 |      |    |      |           |          |           U from PUCT formula,
 |      |    |      |           |          |           see below.
 |      |    |      |           |         Average value of V in a subtree
 |      |    |      |          Probability of this move, from NN, but if Dirichlet
 |      |    |      |          node is on, it's also added here, 0%..100%
 |      |    |     Expected outcome for this position, directly from NN, -100%..100%
 |      |   How many visits are processed by other threads when this is printed.
 |     Number of visits. The move with maximum visits is chosen for play.
Move

* U = P * Cpuct * sqrt(sum of N of all moves) / (N + 1)
  CPuct is a search parameter, can be changed with a command line flag.
* The move with largest Q+U will be visited next
```

**Comment by chara1ampos** [comment]
I think that brute force engines like Stockfish and Houdini have the advantage that their evaluation is cheap, and they can search very deep, thus having great tactics.

Leela's evaluation is very expensive, and thus she cannot search deep enough to avoid blunders. I sense that if one could speed up her evaluation, so she could search deeper, her blunders would be greatly reduced.

On an Nvidia Titan V, where Leela cudnn can evaluate 8000 nodes per second, _she did not seem to blunder_, and even _won several games against Stockfish, Komodo and Houdini_:
[https://groups.google.com/forum/#!topic/lczero/YFMOPQ-J-q4](url)

I recall that alpha zero evaluated around 100000 nodes per second on the deep mind supercomputer, which greatly improves its tactics. This begs the question: what nps did alpha zero use during its training process? I suspect this can greatly affect the quality of the games during training.

**Comment by mooskagh** [comment]
I've added a form for problematic positions submission into the original message.
Sorry for bureaucracy, but that makes it much easier to see the problem.

**Comment by Ishinoshita** [comment]
@chara1ampos : The DM paper says "During training, each MCTS used 800 simulations.", which is a bit ambiguous and may read as new playouts added to the tree or as visits for selected node. Thus nps is irrelevant (but for the total training time). 800 'simulations' is anyway far below 10K's of simulations you mention for match games. So, yes, AZC training may have included blunders as well, at least in early stages (like where we stand now).

**Comment by Why-Sensei** [comment]
ID: 0001
Game: https://lichess.org/FI3y76b0
Bad move: 22. Ng4 (and 21. Rxg8+ wasn't optimal either IMHO)
Correct move: 22. Nf1
Screenshot:
![grafik](https://user-images.githubusercontent.com/37845287/39809692-cdd3f10c-5382-11e8-9917-7d5699cdf86c.png)
Configuration:
Game: LCZero cuDNN 20180508 ID 263 vs Stockfish 9 18050811 Lvl 20
GUI: Arena 3.5.1
Settings: Threads/Cores: 4, Hash Memory: 64MB, Table Memory: 0MB, Ponder: Off, Own Book: Off, Book/Position Learning: Off, Book: none, EGTB: none
System: Win 10 Professional, NVIDIA GTX 1070, 64 GB RAM
Network ID: 263
Time control: 40/12 (adjusted to CCRL)
Comments: Game was streamed on May 9th at https://www.twitch.tv/y_sensei

**Comment by hsntgm** [comment]
@chara1ampos why anybody ask this question maybe Alpha zero just a auto tuned stockfish derivative with neural network.The traditional chess engines elo depends tuning parameters in their code.Maybe they just do that in neural network.

Stockfish 1.01 elo 2754 in 2008
Stockfish 9 elo 3444 in 2018

Look stockfish development history it gained only 700 elo in ten years with billion cpu time and genius c programmers whose tuned parameters step by step.Now we wait Leela gains 500 elo with self playing.Who knows maybe the road map is totally wrong.

Why i think that because someone says leela draws with stockfish ok very good news but how can you explain these blunders and tactical weakness 3000 elo program? Leela's skeleton formed after 10 million game there is no return and this is big paradox for project.

**Comment by Why-Sensei** [comment]
* ID: 0002
* Game: https://lichess.org/efi0R82j
* Bad move: 21. Qxg7
* Correct move: 21. Re3
* Screenshot:
![grafik](https://user-images.githubusercontent.com/37845287/39865749-301305da-544e-11e8-9552-13ad68795270.png)
* Configuration:
Game: LCZero 0.9 ID 271 vs Stockfish 9 18050909 Lvl 20
GUI: Arena 3.5.1
Settings: Threads/Cores: 4, Hash Memory: 64MB, Table Memory: 0MB, Ponder: Off, Own Book: Off, Book/Position Learning: Off, Book: none, EGTB: none
System: Win 10 Professional, NVIDIA GTX 1070, Ryzen 7 1800X, 64 GB RAM
* Network ID: 271
* Time control: 40/4 (adjusted to CCRL)
* Comments: Game was streamed on May 10th at https://www.twitch.tv/y_sensei

**Comment by mooskagh** [comment]
Thanks for submitted the bug reports, they were very useful.

All the blunders so far can be explained by #576.
The fix is there in client v0.10, but it will take multiple network generations to recover the network.

So for a few days (until at ~300000-500000 games are generated by v0.10 client and network is trained on that), don't submit any other positions, as they are likely caused by the same bug.

After that new blunder reports are very welcome!

**Comment by mooskagh** [comment]
For now it would be the most interesting to see examples of blunders that appeared recently.
E.g. if LCzero played correct move in network id270 and now blunders.
That way we'd have some examples of what exactly it unlearns and could look into training data.

**Comment by marcmaths** [comment]
ID: ID288CCLSGame65
Game: https://lichess.org/8mCbbkwl#240
Bad move:  121. Rc7.
Correct Move: Many other moves
Screenshot:
![image](https://user-images.githubusercontent.com/7502112/40013090-e0b8678e-57ac-11e8-8d67-8d8095fbaedb.png)
Configuration:  CCLS Gauntlet (TC 1/1)
Network ID: 288
Time control: 1 min + 1 sec (increment)
Comment:  Game was streamed on May 14th 2018.

**Comment by TCECfan** [comment]
ID: ID288CCLSGame53
Game:  https://lichess.org/0YJMfRI6#260 
Bad move: 131. Ra6.
Correct Move: 131. Ba1
Screenshot: 
![image](https://user-images.githubusercontent.com/7502112/40013904-3ae430d8-57af-11e8-9365-44b5022bed03.png)
![image](https://user-images.githubusercontent.com/7502112/40014046-aea43180-57af-11e8-8017-3dbb3ee4e8a8.png)
Configuration: CCLS Gauntlet 
Network ID: 288
Time control: 1 min + 1 sec (increment)
Comment: Game was streamed on May 14th 2018. Analysis by ID288 in Arena on my machine is also shown above.

**Comment by TCECfan** [comment]
ID: ID288CCLSGame72
Game: https://lichess.org/CVYOwXSK 
Bad Evaluation:  Drew by 3-fold repetition with an evaluation of +15.98
Screenshot:
![image](https://user-images.githubusercontent.com/7502112/40014983-4fa4c458-57b2-11e8-80a9-18db74f04dd8.png)
Configuration: CCLS Gauntlet
Network ID: 288
Time control: 1 min + 1 sec (increment)
Comment: Game was streamed on May 14th 2018.

**Comment by TCECfan** [comment]
There I lots more examples but I will stop here then :)

**Comment by TCECfan** [comment]
I couldn't resist one more...
ID: ID280CCLSGame7
Game:  https://lichess.org/rWWqu4tx#98 
Rh7 would end the game immediately by 3-fold repetition, but Leela played the losing move Kh3 instead: 
![image](https://user-images.githubusercontent.com/7502112/40016943-37400d7c-57b8-11e8-8a7b-a40c503eb074.png)
Configuration: CCLS Gauntlet
Network ID: 280
Time control: 1 min + 1 sec (increment)
Comment:  Does Leela handle 3-fold repetition correctly?

**Comment by apleasantillusion** [comment]
Interestingly, on the Rc7?? Kxc7 and Ra6 Kxa6?? blunders above, I can reproduce them with ID288 on CPU both with game history. 

With just FEN, while it doesn't play both blunders, the killing responses to both blunders are given very low probability from policy, so it's just dumb luck that the engine doesn't play the blunder.

The really interesting part is that with the FEN modified so the 50-move rule halfmove counter set to 0, it immediately sees both killing moves with very high policy outputs.

This is also true of this recent match game: http://lczero.org/match_game/268131

With game history or FEN, 292 plays Rc7??, giving the obvious capture as a response very, very low policy output. 

With FEN altered so 50-move rule halfmove counter set to 0, it immediately sees the capture with 99% probability from policy.

Maybe these examples are just lucky, but it seems high counts for the 50-move rule counter halfmove clock correlate with very strange blunders.

**Comment by nelsongribeiro** [comment]
http://lczero.org/match_game/268155

ID 292 blunders again against ID 233 near the 50-move rule coming up...

**Comment by trophymursky** [comment]
interesting bit based off of apleasantillusion's comment (tho I'm using 292). 

the fen for the interesting position is "2r5/R7/8/8/5k2/8/2K5/8 w - - 85 121" where the policy net ID292 has Rc7 (wrongfully) at 99.91%. 

specifically if you set it to 60 half moves (instead of 85) the policy net fro Rc7 is at .07%. At 65 half moves it's at .2%, at 66 it's at .71%, 67 it's at 1.23%, 68 it's at 6.53% (no longer considered the worst move), 69 it's at 89.47 percent. 

I have no idea why the inflection point would be anywhere near where it, but it's definitely interesting and points towards a training bug corrupting the policy net.

**Comment by so-much-meta** [comment]
FYI... Regarding the a7c7 rook blunder above, I think this might be explained by https://github.com/glinscott/leela-chess/issues/607

Network 288..

With history:
position fen 7r/8/R7/3k4/8/8/2K5/8 w - - 77 117 moves a6a5 d5e6 a5a6 e6f5 a6a5 f5f4 a5a7 h8c8
go nodes 1000
(==> This chooses Kb3)
info string   Kb2 ->       0   (V: 59.51%) (N:  0.29%) PV: Kb2 
info string   Kd1 ->       0   (V: 59.51%) (N:  0.90%) PV: Kd1 
info string   Kb1 ->       2   (V: 52.08%) (N:  1.86%) PV: Kb1 Ke3 Ra3+
info string   Kd2 ->       5   (V: 59.96%) (N:  2.27%) PV: Kd2 Kf5 Rb7 Kf4
info string   Kd3 ->      11   (V: 60.53%) (N:  9.70%) PV: Kd3 Ke5 Re7+ Kd6 Re8 Kd5
info string   Rc7 ->     381   (V: 67.65%) (N: 80.40%) PV: Rc7 Rb8 Rb7 Kf5 Rxb8 Ke6 Kd3 Kd5 Rb5+ Kc6 Kc4
info string   Kb3 ->     491   (V: 83.56%) (N:  4.58%) PV: Kb3 Ke5 Rc7 Kd6 Rxc8 Kd7 Rc5 Kd6 Kc4 Ke6
info string stm White winrate 76.24%


Without history:
position fen 2r5/R7/8/8/5k2/8/2K5/8 w - - 85 121
go nodes 1000
(==> This chooses Rc7)
info string   Kd1 ->       0   (V: 61.53%) (N:  0.00%) PV: Kd1 
info string   Kd2 ->       0   (V: 61.53%) (N:  0.00%) PV: Kd2 
info string   Kb2 ->       0   (V: 61.53%) (N:  0.00%) PV: Kb2 
info string   Kb3 ->       0   (V: 61.53%) (N:  0.00%) PV: Kb3 
info string   Kd3 ->       0   (V: 61.53%) (N:  0.00%) PV: Kd3 
info string   Kb1 ->       0   (V: 61.53%) (N:  0.01%) PV: Kb1 
info string   Rc7 ->     500   (V: 70.63%) (N: 99.98%) PV: Rc7 Rb8 Rb7 Kf5 Rxb8 Ke6 Kd3 Kd5 Rb5+ Kc6 Kc4

**Comment by so-much-meta** [comment]
As to the a7c7 blunder above, I think the history's only part of the problem... The other part of the issue is that the All Ones plane (last input plane) bug really messed up policies.

Good input data was being trained on a bad policy. Consider the effect of the negative log loss/cross entropy in these examples (non-buggy network with low outputs getting trained on a buggy high output).

Here's output from network ID 280. Notice that the a7c7 move only has high probability when the all ones input plane was buggy. Essentially, I think it was bad data like this that kept messing things up.

**History + AllOnesBug**
Policy
              **('a7c7', 0.8687417),**
              ('c2d3', 0.046122313),
              ('c2b3', 0.034792475),
              ('c2d2', 0.03021726),
              ('c2d1', 0.0111367665),
              ('c2b1', 0.006555821),
              ('c2b2', 0.0024336604),
Value: 0.5331184417009354

**History + NoBug**
              ('c2d3', 0.47858498),
              ('c2b3', 0.13757008),
              ('c2d2', 0.13545689),
              ('c2d1', 0.08749167),
              ('c2b1', 0.08396132),
              ('c2b2', 0.07649834),
              **('a7c7', 0.000436759),**
Value: 0.5014338248874992


**NoHistory + AllOnesBug**
Policy:
              **('a7c7', 0.99920577),**
              ('c2d2', 0.00019510729),
              ('c2b3', 0.00015975242),
              ('c2d3', 0.00015850786),
              ('c2b1', 0.0001421545),
              ('c2d1', 7.9948644e-05),
              ('c2b2', 5.882576e-05)]),
Value: 0.5555554553866386

**NoHistory+NoBug**
             ('c2d3', 0.34282845),
              ('c2b3', 0.22524531),
              ('c2d2', 0.14119184),
              ('c2b2', 0.09196934),
              ('c2d1', 0.09108826),
              ('c2b1', 0.08420463),
              **('a7c7', 0.023472117),**
Value: 0.49658756237477064


Now look how all of that changed by network 286, below - now the input with missing history is starting to show the bad policy:

**History+AllOnesBug**
             **('a7c7', 0.88481957),**
              ('c2d3', 0.043222357),
              ('c2d2', 0.030274319),
              ('c2b3', 0.017787572),
              ('c2b1', 0.011131173),
              ('c2b2', 0.011077223),
              ('c2d1', 0.0016878309),
 0.8049132525920868)

**History+NoBug**
(OrderedDict([('c2d3', 0.35683072),
              ('c2b3', 0.17884524),
              ('c2d2', 0.15325584),
              ('c2b2', 0.1069537),
              ('c2d1', 0.10222348),
              ('c2b1', 0.10148263),
              **('a7c7', 0.00040832962)]),**
 0.5084156421944499)

**NoHistory+AllOnesBug**
             **('a7c7', 0.9984926),**
              ('c2d3', 0.00064814655),
              ('c2b1', 0.00030561475),
              ('c2d2', 0.00022950297),
              ('c2b3', 0.00016663132),
              ('c2d1', 8.821991e-05),
              ('c2b2', 6.930062e-05)]),
 0.8271850347518921)

**NoHistory+NoBug**
             ('c2b3', 0.35689142),
              **('a7c7', 0.227083),**
              ('c2d2', 0.1410887),
              ('c2d3', 0.10505199),
              ('c2b1', 0.078001626),
              ('c2d1', 0.0670605),
              ('c2b2', 0.024822742)]),
 0.49565275525674224)


By the time it got to network 288, the policy was really bad in this particular spot:
History+AllOnesBug
             **('a7c7', 0.81777406),**
              ('c2b1', 0.0735284),
              ('c2d3', 0.045673266),
              ('c2d2', 0.044812158),
              ('c2d1', 0.011020878),
              ('c2b3', 0.0059179077),
              ('c2b2', 0.0012732706),
 0.9999993741512299)

History+NoBug
             **('a7c7', 0.8040016),**
              ('c2d3', 0.0970014),
              ('c2b3', 0.04580218),
              ('c2d2', 0.022658937),
              ('c2b1', 0.018647738),
              ('c2d1', 0.008990083),
              ('c2b2', 0.0028980032),
 0.5951071679592133

NoHistory+AllOnesBug
             ('c2b1', 0.30733383),
              **('a7c7', 0.25477663),**
              ('c2d2', 0.19509505),
              ('c2d3', 0.17735933),
              ('c2d1', 0.037348717),
              ('c2b3', 0.02388807),
              ('c2b2', 0.004198352),
 0.9999998211860657

NoHistory+NoBug
             **('a7c7', 0.99980253),**
              ('c2b1', 6.103614e-05),
              ('c2d3', 4.706335e-05),
              ('c2b3', 3.6989695e-05),
              ('c2b2', 2.2621784e-05),
              ('c2d2', 1.6375083e-05),
              ('c2d1', 1.3423687e-05),
 0.6152948960661888



Now, at network 294, this is the current situation:
**History+NoBug**
              ('c2d3', 0.32457772),
              ('c2b1', 0.19262017),
              ('c2d1', 0.15003791),
              ('c2b3', 0.12282815),
              ('c2d2', 0.10260171),
              ('c2b2', 0.08874603),
              **('a7c7', 0.018588383),**
 0.46542854234576225)

**NoHistory+NoBug**
               **('a7c7', 0.99916804)**,
              ('c2b1', 0.00017883514),
              ('c2d1', 0.00016860983),
              ('c2d3', 0.00016126267),
              ('c2b2', 0.00012590773),
              ('c2d2', 0.00010842814),
              ('c2b3', 8.8898094e-05)]),
 0.43435238301754)

**Comment by gyathaar** [comment]
Does it still blunder in those positions if you use --fpu_reduction=0.01  (instead of default 0.1) ?

**Comment by apleasantillusion** [comment]
In the game nelsongribeiro posted, the same pattern holds true (tested with 292).

With history, she plays 124.Ke7 with a very high probability from policy (84.89%), and the response Qxd5 just taking the hanging queen is given only a 2.93% from policy.

Without history at the root, just FEN, she again plays Ke7 with high probability from policy (95.83%), and the Qxd5 response taking the hanging queen is given only 2.33% from policy.

With the FEN modified in only one way, setting 50-move rule counter to 0, Ke7's policy drops to 37.34%, and Qxd5 after Ke7 jumps to 95.07%

Now, from a purely objective standpoint in this particular position, none of this matters so much, since the position is losing to begin with, although forcing black to find the winning idea in the king and pawn ending is a much stronger way of playing than just hanging the queen.

Also, independently of that, the fact that taking a hanging queen is only ~2% from policy when the 50-move rule counter is high is a bit disturbing and is in line with the other examples I cited above.

In general, the variation in probability for Qxd5 based on the 50-move rule counter is quite odd.

In that exact position with black to move (6q1/4K2k/6p1/3Q1p1p/7P/6P1/8/8 b - - 0 0), here are probabilities for Qxd5 with different values of 50-move rule counter:

0:    68.26%
1:    76.71%
5:    89.40%
10:  91.63%
20:  92.48%
30:  94.28%
40:  89.57%
50:  77.83%
60:  83.95%
70:  52.39%
80:  66.84%
90:  11.43%
99:   1.06%

**Comment by nelsongribeiro** [comment]
The really bad move is the move is just before that position:
The FEN position is (4K3/6qk/3Q2p1/5p1p/7P/6P1/8/8 w - - 94 123). 

Its a draw at this point, on move 123, ID 292  played _123. QD5_ intead of 123. QE5.

Last time that a pawn was moved was at 75...f5 , what makes this the 49th move after that.

**Comment by apleasantillusion** [comment]
Just to add to this, the Rxf3+ that's being tracked in the sheet at https://docs.google.com/spreadsheets/d/1884-iHTzR73AgFm19YYymg2yKnwhtHdaGyHLVnffLag/edit#gid=0 shows the same behavior. 

With net 297, probability with various 50-move rule values:

0: 0.07%
10: 1.88%
25: 10.23%
50: 1.11%
90: 1.72%
99: 6.97%

That's some heavy variation just from changing the 50-move rule counter.

Also, the pattern is different with this one. In all the others, probabilities were worst at the very high counts, a bit better at very low counts, and best at counts around 30. Here that last trend maintains, but the other is more muddled.

**Comment by ASilver** [comment]
I don't know if it is a consequence of the bug, or the new PUCT values which inhibit tactics, but the latest versions (I am watching 303 right now) have some appallingly weak ideas of king safety and closed positions. I am playing a match against id223 at 1m+1s and it is more than a tactical weakness issue, it is one of completely wrong evaluations, which 223 did not have, that is leading it to happily let its king be attacked until it is too late to save itself. I also saw more than one case where it thought a dead drawn blocked game, with no entry or pieces, was +2, while 223 thought it about equal. The result was that 303 preferred to sacrifice a pawn or two to not allow a draw, and then lost quickly thanks to the material it gave away.

**Comment by TCECfan** [comment]
@ASilver  I have watched many games from many of the CCLS gauntlets, and my overall view (from the perspective of spectator) is that her style changed markedly following the release of v0.8.  In particular, she started showing: 
Type 1:  unstable or poor evaluations, and inflexible play.   
Type 2: "buggy-looking" play in closed positions in which both engines are shuffling, in repeated positions (especially 3-fold), and in positions where the 50-move rule is key.
In my view, Type 1 has "followed the trajectory of the value head", whereas Type 2 has seemed to persist even as "the value head has partially recovered".

**Comment by so-much-meta** [comment]
Regarding the rook blunder above:
position fen 7r/8/R7/3k4/8/8/2K5/8 w - - 77 117 moves a6a5 d5e6 a5a6 e6f5 a6a5 f5f4 a5a7 h8c8
(or just "2r5/R7/8/8/5k2/8/2K5/8 w - - 85 121")

I did some more analysis. I think some here might find it useful. Check out the attached PDFs with graphs.

What I did was I iterated the halfmove clock (rule50) from 0 through 91, created the a FEN position at each halfmove clock, played each of the 8 moves, and determined the network value and policy under the following conditions:
(1) Normal History
(2) Fake History [All history planes are just copied from the first]
(3) No History [All history planes are set to 0]

I did this with both net 288 and net 303. So that's 12 graphs. (halfmove clock is horizontal axis).

I also added the move a7c7 (the blunder move - where Leela does not want to take the free rook, so blunders again in the PV), and did the exact same process.

Some findings:
* The blunder is strongly correlated with the halfmove clock.
* The situation is clearly getting better going from 288 up to 303
* Fake history (as opposed to all zeroes) would be very advantageous here, if no history is available.


[rook_blunder_Id288_Id303.pdf](https://github.com/glinscott/leela-chess/files/2012051/rook_blunder_Id288_Id303.pdf)
[rook_blunder_Id288_Id303_a7c7.pdf](https://github.com/glinscott/leela-chess/files/2012052/rook_blunder_Id288_Id303_a7c7.pdf)

**Comment by peterjacobi** [comment]
Good examples for the longstanding problem with discovered attacks. Furunkel just posted in #game-analysis, but should be preserved here:

> Re-looking at the recent ID300 gauntlet, discoveries are still a huge problem for Leela. In the 
> first 50 games, Leela fell 6 times for a discovered attack. This was not always a game loser but 
> still. I collected those 6 games if people want to use these for testing: [https://pastebin.com/mffqHESV](https://pastebin.com/mffqHESV)
> The critical moves in those were: game 1: 28.Rb7, game 2:23.Nd5, game 3:  35.Qe4 (ouch), 
> game 4: 29.Rh6 (ouch), game 5: 30.Nc5, game 6: 15..Qd4 (ouch). The 3 I labeled "ouch" are 
> particularly bad.

Also attached: [leela300-discoveries.txt](https://github.com/glinscott/leela-chess/files/2012418/leela300-discoveries.txt)

**Comment by hsntgm** [comment]
Latest ID's have higher elo so i tested ID 303 with 60 min game with stockfish yesterday i watched a whole game.

I was hoping a good result but it seems she still suffer this unknown bug in deep.
I am not strong player just 1700 and i simply see especially her position evaluation and simple tactic understanding terribly broken if i compare older networks.So this elo gains is not true.

It seems it will be very difficult to get rid of the past bug effect.She has broken heart.If you believe that you found the bug in algorithm and corrected it maybe you just have to start whole training process again.If the problem is training data i think you have to start again because she will have to do two times more training than is necessary to get rid of this effect.

**Comment by TCECfan** [comment]
I am just posting here an example that was posted on Discord #technical three hours ago.  It is yet another awful 50-move-rule related blunder (the 50-move count starts at move 48 in this example).  The blunder is made on move 96 of this game: https://lichess.org/Y1zkwwuv#190

**Comment by steve3140** [comment]
Haven't posted here before, but noticed this blunder in a match game.  Candidate ID 306 was playing Black in this game:

http://lczero.org/match_game/281647

and blundered into a discovered check with 60. ... Ke6 and dropped a bishop, so the game petered out to a draw.

**Comment by so-much-meta** [comment]
Alexander - good point about the graph scale. I recognized that, of course, but just kinda had a late night "data is data" attitude... This is the KRkr position where Leela tries to block a rook check with an unprotected rook.

Correct first move is mostly anything but a7c7, the move where Leela tries to give away its rook in net 288 (and only after ~75 halfmoves). But c2d3 is best, moving the king toward the center. (both nets want to do that when halfmoves aren't too big).

Correct second move is c8c7 (taking the unprotected rook). Both ID288 and ID303 think that's a bad idea when halfmoves is big.

In all cases, not having history amplifies the error.

**Comment by haleysa** [comment]
I have been tracking 3 of the earliest 1-move tactics misses in this thread for the last few nets, and also looked into them to see what's going on under the covers a bit.

Position 1 - The missed pin  https://lichess.org/e07JvP6g#43
Leela is totally blind to the mate in 1 threat, presumably because she thinks the mating square is covered by the d2 rook, but the rook is pinned.  The problem is not the policy head on the blunder move; actually, Leela's policy prefers the #1 SF move of Kb1.  However, the mating move Qxc2# has such a low policy on any of the many blunder moves Leela might make, it takes 30000+ nodes down each "blunder move" to realize they are bad and start preferring a new move.  More search won't solve this easily; it would take a lot of nodes.  At 100k Leela still blunders, just with a different move Nb6 or Qxb4.
The policy for Qxc2# has gone from 0.13% at ID297 to 0.21% at ID308. It's slow improvement, if at all.

Position 2 - Discovered attack with check https://lichess.org/mbWjiT93#105
Leela is blind to the threat that was created of R6g+ and a discovered attack on the black queen.  In ID297 at 10k nodes Leela prefers a different losing move, b4; but by ID302 at 10k the preferred move is the SF best move of Qf3 - close enough that it's searchable at a reasonable time control.  The policy for the refutation move Rg6+ has gone from 0.43% at ID297 to 0.39% at ID308.  It's gotten as high as 0.70% in ID302; ID308 was a downswing.

Position 3 - Missed mate threat after capture https://lichess.org/FI3y76b0#43
Leela needs to stop mate threat of Qh2.  Both Ng4 and Nf1 guard the h2 square, but Ng4 allows the knight to be captured by the rook.  I believe Leela puts this as a low probability because the knight is guarded, so it would be dropping the exchange - except of course it's dropping mate elsewhere instead.
Here the policy for Ng4 is very high - about 90% - and the policy for the Rxg4 refutation is very low, 0.09 for ID297 up to 0.14 for ID308.  10k nodes is not enough to prevent the blunder or find the refutation, but 100k nodes is.

Also in this thread that I found interesting, but haven't tracked closely, is a 2-move tactic where Leela drops the queen:  https://lichess.org/efi0R82j#43
Leela takes the pawn with Qxg7 and her queen gets trapped with Rg6.  It's actually a miss of both position 2's type and position 3's type - the two ways her queen could try to get out are Qxh7; this is met with a discovered attack on the queen with check of Rg2+; and Rxe5 threatening the black queen; this is met with Rxg7 and if white responds by taking Rxe6 then black has Rxe1# (this is the path taken in the game, she avoids mate but drops the queen).  Being 2 ply deeper in search, and with two "good" moves that have a hard to spot refutation, this takes a huge amount of search to avoid (more than 100k and I suspect a lot more).  But this should go away once both of the other issues are learned.

**Comment by rwbc** [comment]
@haleysa
good summary!

**Comment by steve3140** [comment]
If this would benefit from an ID, let's call it **MultiMove#1**.

I'm also interested in multi-move tactics and noticed one when I was playing some matches between NN226 and NN300 last night.  It involves a couple of themes where I've noticed Leela making some bad mistakes recently, which in this example are:

* theme: removal of defender (capturing a piece which defended the queen, leaving the queen hanging); combined with
* theme: discovered check (knight gives king check, revealing discovered attack against the now-hanging queen)

Here's the moves of the game:
LC Zero 226 GPU (2800) - LC Zero 300 GPU (2800) [B51]
1.e4 c5 2.Nf3 d6 3.Bb5+ Nd7 4.Bxd7+ Bxd7 5.d3 e6 6.0-0 Ne7 7.d4 Ng6 8.dxc5 dxc5 9.Nc3 Be7 10.Qe2 Qc7 11.Rd1 0-0 12.Nb5 Bxb5 13.Qxb5 Rfd8 14.Bd2 a6 15.Qe2 b5 16.g3 Qb7 17.h4 Nf8 18.b3 b4 19.c4 a5 20.a4 bxa3 21.Rxa3 Ra7 22.Rda1 Rda8 23.Kg2 Nd7 24.Bc3 Nb8 25.Rxa5 Rxa5 26.Rxa5 Rxa5 27.Bxa5 Qxb3 28.e5 Qb7 29.Bc3 Na6 30.Kh2 Nb4 31.Nd2 h6 32.Ne4 Qd7 33.Qe3 Qc6 34.Qf4 Nd3 35.Qf3 Nb4 36.Bxb4 cxb4 37.Nf6+ gxf6 38.Qxc6 fxe5 39.Qe8+ Bf8 40.c5 b3 41.c6 b2 42.c7 b1Q 43.c8Q Qb4 44.Qcd7 Qe7 1-0

Time control was Blitz 2m+1sec/move.  All engine settings were defaults.

The tactics are at move 35 where NN226 playing White lays a 'trap' with 35.Qf3 and NN300 as Black stumbles into it with 35...Nb4 (best/saving move = 35...Qd7).

FEN for position just before Nb4 is:  6k1/4bpp1/2q1p2p/2p1P3/2P1N2P/2Bn1QP1/5P1K/8 b - - 0 35

![1](https://user-images.githubusercontent.com/39388181/40257677-6e3009fe-5b32-11e8-8a07-83935621a635.jpg)

Checked some other versions of Leela in Chessbase Reader as an analysis engine: 
NN253 finds 35...Nb4 problematic after 6kN and eventually recommends 35...Qd7 after 16kN.  
**NN311** recommends 35...Nb4 and doesn't see any problems until 29kN have been searched (on my hardware that's 4 mins).
NN311 eventually locates 35...Qd7 after 59kN (7mins50sec).
After 35...Nb4 is played (as per game), NN311 recommends 36.Qg4 for White and persists with this recommendation until I gave up the search after about 10 mins.
After 36.Bxb4 is played (as per game), NN311 then sees the tactic within 1kN and moves the queen away.

**Comment by steve3140** [comment]
So I finally spent a few minutes learning about the CLI after seeing someone else's post, so I've worked out how to get more detailed info.  Here's a comparison of that MultiMove#1 position from a few different versions: 

lczero.exe -w <filename>
position fen 6k1/4bpp1/2q1p2p/2p1P3/2P1N2P/2Bn1QP1/5P1K/8 b - - 0 35
go movetime 1000

NN253
info string   Qd7 ->      15   (V: 34.55%) (N: 27.76%) PV: Qd7 Nd6 Bxd6 Qxd3 Qb7 exd6
info string   Nb4 ->      81   (V: 45.14%) (N: 40.43%) PV: Nb4 Qg4 Nd3 Qe2 Nb4 Qg4 Nd3 Qe2
info string stm Black winrate 42.47%

NN292
info string   Qd7 ->      22   (V: 33.95%) (N: 33.45%) PV: Qd7 Nd6 Nxe5 Bxe5 Bxd6 Qd3 Qc6 Qxd6 Qf3
info string   Nb4 ->      67   (V: 45.35%) (N: 20.13%) PV: Nb4 Qg4 Nd3 Nf6+ Kh8 Nh5 Bf8
info string stm Black winrate 40.59%

NN311
info string   Qd7 ->      19   (V: 38.07%) (N: 40.24%) PV: Qd7 Nd6 Nxe5 Bxe5 Bxd6 Qd3 Qc6
info string   Nb4 ->      41   (V: 49.85%) (N: 24.53%) PV: Nb4 h5 Qa6 Nd6 Bxd6 exd6 Qxd6 Qa8+
info string stm Black winrate 44.86%

NN316
info string   Qd7 ->      17   (V: 32.87%) (N: 41.50%) PV: Qd7 Nd6 Nxe5 Bxe5 Bxd6 Qd3 Qc6 Qxd6
info string   Nb4 ->      77   (V: 48.53%) (N: 21.11%) PV: Nb4 Qe3 Nc2 Qd3 Nb4 Qe2 Qa6 Nd6
info string stm Black winrate 44.36%

At a longer 1 minute (movetime 60000) search, NN316 still recommends the losing move Nb4 after 1600 nodes:
info string   Qd7 ->      69   (V: 21.95%) (N: 41.50%) PV: Qd7 Nd6 Nxe5 Bxe5 Bxd6 Qd3 Qc6 Qxd6 Qf3 Qxc5 f6 Bd4 e5 Be3 Kh7
info string   Nb4 ->    4014   (V: 44.92%) (N: 21.11%) PV: Nb4 Qg4 Kh8 h5 Bf8 Nd6 Kg8 Qe4 Qd7 Qa8 Nd3 f4 Nf2
info string stm Black winrate 44.21%

With a 10 minute (movetime 600000) search, NN316 finds the best move Qd7 after 24000 nodes:
info string   Nb4 ->    9609   (V: 35.46%) (N: 21.11%) PV: Nb4 Qg4 Kf8 Nd6 Kg8 Ne4 Kf8 Nd6 Kg8 Ne4
info string   Qd7 ->   46166   (V: 35.93%) (N: 41.50%) PV: Qd7 Nd6 Nxe5 Bxe5 f6 Qa8+ Bf8 f4 h5 Kh3 fxe5 fxe5 g6 Kg2 Kg7 Qa6 Kh7 Qb6 Bg7 Qxc5 Qa4 Qd4 Qc2+ Kf3 Qb3+ Kf2 Qc2+ Ke1 Qc1+ Ke2 Qc2+ Kf1 Qc1+ Kg2 Qc2+ Kf3 Qc1 Ne4 Qf1+ Ke3 Qg1+
info string stm Black winrate 35.77%

Question (from one who is ignorant):  Given a 41.50% vs 21.11% differential, why was Qd7 visited so little to start with?

**Comment by mooskagh** [comment]
@steve3140 It means that after trying to make that move, the output of value head was so low that it decided that prior is wrong.

Also it's known to have especially bad result when no-capture counter is around 30-40, and when there is no history in fen position. It seems that in your tests you had both.

**Comment by ASilver** [comment]
What if the PUCT and FPU values were changed? Have you tested?

**Comment by haleysa** [comment]
No, but I don't see much reason to test - of course modifying PUCT and FPU will help, finding the refutation moves is a matter of searching for a single move that immediately changes the valuation significantly.  The first refutation position is mate in 1; as soon as the search starts touching the position, it elevates it to best move pretty much immediately.  But the policy head is so low in some nets it's basically tied for last choice of moves.  Tuning PUCT/FPU to make sure we touch all moves in under 1000 nodes is not a usable solution, the policy head needs to know these refutations are good moves.

**Comment by so-much-meta** [comment]
Reverting PUCT/FPU might help in training (to get the policy head to see
the correct move so that it's correctly trained in the first place)

**Comment by ASilver** [comment]
You don't need to touch all moves, but there is a difference between touching all and not enough. What if widening it led it to find more correct moves, and thus learn from them, and thus... improve?

**Comment by ASilver** [comment]
As reported in the Discord channel, in this game, on move 138, NN303 hangs a rook out of the blue costing the drawn game almost instantly. The next move the eval had swung +9.42:

[Event "DESKTOP-RV5DCNB, Blitz 1m+1s"]
[Site "Rio de Janeiro, Brazil"]
[Date "2018.05.18"]
[Round "18"]
[White "Spike 1.4"]
[Black "lczero v0.10"]
[Result "1-0"]
[ECO "D94"]
[Annotator "0.09;0.01"]
[PlyCount "283"]
[TimeControl "60+1"]

{Intel(R) Core(TM) i5-2500K CPU @ 3.30GHz 3293 MHz W=22.2 plies; 4,701kN/s
B=17.1 plies; 1kN/s; 45 TBAs} 1. c4 c5 2. Nf3 Nf6 3. Nc3 e6 4. e3 Nc6 5. d4 {
Both last book move} d5 {0.01/18 2} 6. a3 {0.09/16 3 (cxd5)} cxd4 {0.06/18 2 
(a6)} 7. exd4 {0.17/16 2} g6 {-0.02/19 2 (Be7)} 8. Bd3 {0.59/16 2 (cxd5)} Bg7 {
-0.13/19 2 (dxc4)} 9. O-O {0.59/16 2} dxc4 {-0.12/19 2} 10. Bxc4 {0.42/18 12}
O-O {-0.16/19 1} 11. Bg5 {0.47/18 3 (Re1)} h6 {-0.34/18 1} 12. Be3 {0.47/17 3 
(Bh4)} b6 {-0.34/18 2 (Nd5)} 13. Qe2 {0.44/14 2} Bb7 {-0.27/19 2} 14. Rfd1 {
0.36/14 2 (Rad1)} Ne7 {-0.29/18 2} 15. Ba6 {0.34/14 2 (Bd3)} Bxa6 {-0.31/18 2}
16. Qxa6 {0.14/15 3} Nfd5 {-0.32/19 2 (Qc8)} 17. Rac1 {0.09/16 3 (Bd2)} Nxe3 {
-0.43/18 2 (Nf5)} 18. fxe3 {0.45/15 4} Nf5 {-0.42/18 1 (g5)} 19. Qd3 {0.40/15
2 (Qe2)} Rc8 {-0.49/19 2} 20. Rc2 {0.28/15 2 (e4)} Rc7 {-0.60/19 2 (g5)} 21.
Kh1 {0.43/13 2 (Re2)} h5 {-0.71/19 2 (Nd6)} 22. Qe2 {0.22/13 2 (Rcd2)} Nh4 {
-0.62/19 3 (Qb8)} 23. Rdc1 {0.15/15 2 (Nxh4)} Nf5 {-0.79/19 3} 24. Qe1 {
0.09/15 2 (Rd1)} Bh6 {-0.90/19 3} 25. Re2 {0.08/16 3} g5 {-0.66/19 2 (Bg7)} 26.
Rd1 {0.00/15 2} g4 {-0.69/18 1} 27. Ne5 {0.13/17 4} Bg7 {-0.69/19 3 (Qg5)} 28.
Nb5 {0.02/15 2 (Ne4)} Bxe5 {-1.11/18 2} 29. dxe5 {0.00/19 2} Rd7 {-1.06/18 1}
30. Nd6 {0.00/20 1 (Rxd7)} Nxd6 {-0.79/19 3 (Qg5)} 31. Red2 {0.02/18 2} Qg5 {
-0.61/18 1} 32. exd6 {0.00/19 2} Rfd8 {-0.65/20 3 (Qe5)} 33. Rd3 {0.08/19 2 
(h3)} Qe5 {-0.87/18 2 (Qc5)} 34. Qb4 {0.02/17 2 (Qd2)} f6 {-0.83/19 3} 35. Qc4
{0.06/17 2 (Kg1)} f5 {-0.84/19 3 (b5)} 36. Qf4 {0.00/18 1} Qxf4 {-0.85/17 1}
37. exf4 {0.00/23 1} Kf7 {-0.91/19 2} 38. Kg1 {0.00/24 1} Kf6 {-0.92/19 3} 39.
g3 {0.00/25 2 (Kf2)} e5 {-0.82/17 2} 40. fxe5+ {0.00/25 2} Kxe5 {-0.82/18 1}
41. Re3+ {0.00/24 1} Kf6 {-0.75/17 0} 42. Red3 {0.00/27 2} Ke6 {-0.66/19 3 
(Ke5)} 43. Re1+ {0.00/29 2 (Re3+)} Kf6 {-0.64/17 2} 44. Red1 {0.00/10 0} Rc8 {
-0.63/19 3 (Ke5)} 45. R1d2 {0.00/22 1 (Rf1)} Ke6 {-0.57/18 3 (Rc4)} 46. Re3+ {
0.00/23 1 (Kf2)} Kf6 {-0.64/17 2} 47. Kf2 {0.00/25 1} Rcd8 {-0.76/18 1 (Rc6)}
48. Red3 {0.00/26 1} Rh8 {-0.85/18 1 (Ke6)} 49. a4 {0.00/22 1 (Rd5)} h4 {
-1.30/17 2 (Rhd8)} 50. Kg2 {0.00/20 1} h3+ {-1.28/17 2 (hxg3)} 51. Kf2 {
0.00/26 1} Rhd8 {-1.34/18 2} 52. b3 {0.00/29 1 (b4)} Rc8 {-1.45/18 3 (Ke5)} 53.
Ke2 {0.00/28 1 (Ke3)} Rc5 {-1.54/18 3 (Re8+)} 54. Kf2 {0.00/28 1} Rc1 {-1.53/
17 2 (Rc8)} 55. Ke3 {0.00/26 1} Rf1 {-1.54/18 2 (Re1+)} 56. Ke2 {0.00/26 1} Rh1
{-1.50/17 1} 57. Ke3 {0.00/28 2} Rb1 {-1.39/18 2 (Rf1)} 58. Ke2 {0.00/27 1 
(Kf2)} Rh1 {-1.28/17 2} 59. Ke3 {0.00/10 0} a6 {-1.25/18 2 (Re1+)} 60. Rc3 {
0.00/22 1 (Rc2)} Rb1 {-1.38/17 1 (Re1+)} 61. Kf2 {0.00/24 1 (Rcd3)} Rh1 {
-1.27/18 2} 62. Ke3 {0.00/10 0} Re1+ {-1.08/18 2} 63. Kf2 {0.00/30 1} Rb1 {
-1.14/17 1 (Rh1)} 64. Rcd3 {0.00/25 1 (Ke2)} Rh1 {-1.31/17 2 (Ke6)} 65. Ke3 {
0.00/27 1} Rb1 {-1.26/18 2 (Re1+)} 66. Ke2 {0.00/26 1 (Kf2)} Rh1 {-1.30/17 2}
67. Ke3 {0.00/10 0} Rg1 {-1.08/17 1 (Re1+)} 68. Rd5 {0.00/23 1} Rb1 {-1.26/16
1 (Re1+)} 69. R5d3 {0.00/25 1} a5 {-1.32/17 2 (Rg1)} 70. Ke2 {0.00/25 1 (Rc3)}
Rh1 {-1.39/16 1} 71. Ke3 {0.00/27 1} Re1+ {-1.22/16 1 (Rb1)} 72. Kf2 {0.00/32 1
} Rh1 {-0.98/16 1} 73. Ke3 {0.00/10 0} Rg1 {-0.87/17 1 (Re1+)} 74. Rd5 {
0.00/24 1} Re1+ {-0.82/16 1 (Rb1)} 75. Kf2 {0.00/26 1} Re5 {-1.03/16 1 (Rh1)}
76. R5d4 {0.00/26 1} Rc5 {-0.94/17 1 (Ke6)} 77. R4d3 {0.00/28 1} Rc8 {-1.02/17
2 (Rc1)} 78. Ke2 {0.00/29 1 (Ke3)} Re8+ {-0.96/16 1 (Rc5)} 79. Kf2 {0.00/33 1}
Rc8 {-0.90/16 1} 80. Re3 {0.00/28 1 (Ke3)} Rc6 {-1.04/15 1} 81. Red3 {0.00/31 1
} Ke6 {-0.91/16 1 (Rc8)} 82. Re3+ {0.00/31 1} Kf6 {-0.74/16 1} 83. Red3 {
0.00/13 0} Rc5 {-0.74/16 1 (Rc8)} 84. Re3 {0.00/29 1 (Re2)} Rc8 {-0.71/16 1 
(Rc6)} 85. Ke2 {0.00/27 1 (Red3)} Rc6 {-1.16/15 1 (Rcd8)} 86. Red3 {0.00/28 1}
Ke6 {-0.95/17 2 (Rc8)} 87. Re3+ {0.00/31 1} Kf6 {-0.82/17 1} 88. Red3 {0.00/13
0} Rc8 {-0.66/17 1} 89. Re3 {0.00/12 0 (Ke3)} Rcd8 {-0.74/16 1} 90. Red3 {
0.00/30 1} Ke5 {-0.58/17 1 (Re8+)} 91. Re3+ {0.00/10 0} Kf6 {-0.47/14 0} 92.
Red3 {0.00/10 0} Re8+ {-0.38/17 2} 93. Kd1 {0.00/31 1 (Kf2)} Re5 {-0.85/17 1 
(Red8)} 94. Rc2 {0.00/25 1 (Kc2)} Ke6 {-1.32/16 1} 95. Rcd2 {0.00/27 1} b5 {
-1.37/16 1 (Kf6)} 96. axb5 {0.00/25 1} Rxb5 {-1.30/16 1} 97. Ke1 {0.00/26 1 
(Kc2)} Re5+ {-1.20/17 1 (Rb6)} 98. Kf2 {0.00/28 1} Rb5 {-1.00/16 1} 99. Re3+ {
0.00/29 1 (Re2+)} Kf6 {-1.03/16 1 (Re5)} 100. Rd1 {0.00/24 1 (Red3)} Rc5 {
-1.23/16 1} 101. Rd2 {0.00/28 1} Rc1 {-1.13/17 1 (Rb5)} 102. Red3 {0.00/27 1}
Rh1 {-1.09/17 1 (Rc5)} 103. Ke3 {0.00/28 1} Rf1 {-0.94/17 1 (Re1+)} 104. Ke2 {
0.00/26 1} Rh1 {-0.80/16 1} 105. Ke3 {0.00/10 0} Re1+ {-0.70/17 2} 106. Kf2 {
0.00/34 1} Ra1 {-0.72/16 1 (Rh1)} 107. Rd5 {0.00/23 1 (Ke2)} Ke6 {-0.87/16 1}
108. R5d3 {0.00/26 1 (Ke3)} Rh1 {-1.09/16 1 (Kf6)} 109. Ke3 {0.00/27 1} Re1+ {
-1.05/17 1} 110. Kf2 {0.00/28 1} Rh1 {-1.01/17 1 (Re5)} 111. Ke3 {0.00/31 1}
Re1+ {-0.88/17 1} 112. Kf2 {0.00/10 0} Re5 {-0.92/17 1} 113. Rd4 {0.00/28 1 
(Kg1)} Rb5 {-1.03/16 1 (Rc5)} 114. Re2+ {0.00/28 1 (R4d3)} Kf6 {-1.16/16 1 
(Re5)} 115. Rd3 {0.00/27 1} Rb6 {-1.08/16 1 (Re5)} 116. Red2 {0.00/28 1} Rb5 {
-1.01/17 1 (Ke6)} 117. Rd1 {0.00/26 1 (Rd5)} Rb8 {-1.00/16 1 (Rc5)} 118. Re3 {
0.00/26 1 (Ra1)} Rb6 {-1.10/17 2} 119. Red3 {0.00/29 1} Rc6 {-1.07/17 1 (Rb8)}
120. R1d2 {0.00/29 1 (Ke3)} Rb6 {-0.97/16 1 (Rc5)} 121. Kg1 {0.00/29 1 (Rd5)}
Ke6 {-1.17/16 1} 122. Re3+ {0.00/30 2} Kf6 {-0.96/16 1} 123. Red3 {0.00/10 0}
Rb8 {-0.82/17 1 (Ke6)} 124. Kf2 {0.00/31 1} Rb5 {-0.76/16 1 (Rb6)} 125. Rc3 {
0.00/29 1 (Kg1)} Rb6 {-0.73/16 1 (Ke6)} 126. Rcd3 {0.00/10 0} Rc6 {-0.56/16 1 
(Ke6)} 127. Ke3 {0.00/26 2 (Rd5)} Rc5 {-0.54/16 1 (Rb6)} 128. Kd4 {0.00/25 1 
(Kf4)} Rc6 {-0.44/16 1} 129. Ke3 {0.00/10 0} Rc1 {-0.56/17 1 (Rb6)} 130. Ke2 {
0.00/27 1 (Kf2)} Rg1 {-0.72/16 1 (Rh1)} 131. Ke3 {0.00/28 1} Rf1 {-0.65/17 2 
(Re1+)} 132. Ke2 {0.00/26 1} Rg1 {-0.53/16 1} 133. Ke3 {0.00/10 0} Rh1 {
-0.60/17 1 (Re1+)} 134. Rd5 {0.00/25 1} Rb1 {-0.50/16 1} 135. R5d3 {0.00/27 2 
(Rb5)} Kg5 {-0.61/16 1 (Re1+)} 136. Rc2 {0.00/21 2 (Ra2)} Re1+ {-0.51/16 1 
(Kf6)} 137. Kd2 {0.00/22 1 (Kf2)} Rh1 {-0.26/16 1 (Re6)} 138. Kc3 {1.20/16 2}
Rc1 {-0.12/16 1 (Rg1)} 139. Rxc1 {9.30/17 1 (Rd5)} Kf6 {12.90/17 2} 140. Ra1 {
11.08/18 2 (Re1)} Ke6 {12.96/16 1 (Kg5)} 141. Rxa5 {13.27/18 1 (Re1+)} Rd8 {
13.73/16 1 (Kf6)} 142. d7 {19.37/18 2 (Re3+)} 1-0

**Comment by dstark1993** [comment]
Self play No 14200622.
http://lczero.org/game/14200622 
or with game analysis 
https://lichess.org/study/Smk8bomB

18. bxc6 qe8??
Blunder. Best move was Bf5. 
Evaluation from +2.5 to +9.5

![capture](https://user-images.githubusercontent.com/39532052/40380363-8fe689de-5e01-11e8-90e8-9f977cb7d522.PNG)

Network:
1831f4884d6da86fe369d4a51fbfe6a433703fbf97b0e7122898170c1eede5e0
ID329

Its just a game from generated self play using googles colab Leela_K80
Looks like the queen would move to get out of discovery....

**Comment by dstark1993** [comment]
How can you tell if it was randomized move or made by leela "on purpose"? its something that can be checked?

**Comment by so-much-meta** [comment]
@dstark1993 - yeah, you could check this by running the engine outside of training, being sure to provide some history, and seeing what the engine does.... However, that's a little tedious... I'm going to be trying to figure out some good ways to analyze the training data in a more automated way.

**Comment by dstark1993** [comment]
Good, cause i dont understand very much in programming, im generating games with google colab. Dont think ill be able to use lc0 on my laptop (power and/or my skill for figuring how to)

**Comment by chara1ampos** [comment]
Id: Loss of Queen
    Game: [https://lichess.org/fzuBrqhf#72](url)
    Bad move: 36. Qxc2, (Stockfish eval goes from -0.2 to +7.3)
    Correct move: Qc6 
    Configuration: lc0 cudnn - May 19 (default parameters), Windows 10 x64, Nvidia Titan V, Intel i5-7400T quad core, 32 GB RAM
    Network ID: kb1-256x20-2100000.txt.bz2
    Time control: 60" + 2" Time
    Comments: Leela was ahead but blundered, losing a Queen, and then resigned, losing the game to Houdini 1.5.

**Comment by steve3140** [comment]
That last one "Loss of Quuen" is worth monitoring a bit I think.  Its another multi-move blunder with the theme again being "removal of defender".  Am hoping to see Leela gradually work these oversights out of her system.

**Comment by haleysa** [comment]
The "Loss of Queen" removal of defender with check isn't actually too bad - at 10k nodes ID 329 Leela won't play the blunder Qxc2, and while the policy for Qxc2 is N 65.89%, the listed Best Move is the 3rd highest policy at 3.93%.  Interestingly Leela actually prefers Rb8(N 2.21%) at 10k nodes, which isn't a blunder, and Stockfish says the position is even after that.  
After Qxc2, the policy for the refutation Re8+ is 2.17% and is preferred at 5k nodes search.  Rxg5 is also good for white, but not as good (QN v RR for white), and is N 3.78%.  So there's lots of ways for Leela to get out of this.

**Comment by chara1ampos** [comment]
Id: **Loss of Knight**
Game: [https://lichess.org/pJh8VCTA#89](https://lichess.org/pJh8VCTA#89)
Bad move: 45. Rxb5 (Stockfish eval goes from -1 to -5.1)
Correct move: f3
Configuration: lc0.exe cudnn - May 22 (default parameters), Windows 10 x64, Nvidia Titan V, Intel i5-7400T quad core, 32 GB RAM
Network ID: 330
Time control: 60" + 2"
Comments: **Leela blundered, lost a piece, and then resigned.**

**Comment by so-much-meta** [comment]
I've been posting on the forums about some work I'm doing to try to automate blunder detection. Please see here: https://groups.google.com/forum/#!topic/lczero/8lK5ldgZUHA

So far, this does seem to work to find things like buggy (old) engines playing match games, as well as blind spots (missed mate in ones, etc)... However, I haven't fully figured out exactly how I'd like to measure things to get a reliable signal. Once I improve this, I hope to apply this to new match games as they come out to look for bugs and blunders.

**Comment by haleysa** [comment]
I've actually spent a little time today doing something similar with the PGNs from the CCLS gauntlets.  It's easier in a lot of ways because the opposing engine will spot the blunder, so I've started just scanning for situations where Leela's eval drops by a certain threshold (right now -200 centipawns) - this means she made a move and didn't see the refutation that was coming from the other engine.  I'm also filtering it for where Leela's eval didn't start worse than -2.00, because if she's already losing and blunders more it's not so interesting in my opinion.  A lot of that helped avoid "losing faster" endgame moves from cluttering it.
It's still in its infancy and won't be worked on this weekend, but hopefully it can be converted into generating some automatic tracking positions for future IDs as another way to check progress.

**Comment by so-much-meta** [comment]
Using that method, here's the worst I could find from the last 500 match games: 311088-311587

White has a mate in one on move 32, but misses it and gets check-mated a couple moves later:
http://www.lczero.org/match_game/311146

White doesn't protect against black's mate in one on move 46 (according to Lichess, white had a small advantage), but then black misses the opportunity and ends up with a draw:
http://www.lczero.org/match_game/311121

White gives black a mate in two opportunity on both move 94 and 96, but black misses it both times:
http://www.lczero.org/match_game/311554

Black had a forced mate in 5, but ends up with a draw:
http://www.lczero.org/match_game/311240
https://lichess.org/analysis/8/8/pRpB4/2P2k2/P3p2K/4P3/5n2/6r1%20b%20-%20-%204%2047#93

White missed a simple tactic to capture black's queen on move 42, would have won but goes on to lose:
http://www.lczero.org/match_game/311127
https://lichess.org/analysis/6k1/p7/7p/1p1pN3/8/8/1qB2R1K/8%20w%20-%20-%200%2042#82

Black misses trading its rook for a queen on move 38:
http://www.lczero.org/match_game/311379

**Comment by chara1ampos** [comment]
Id: **Leela gets checkmated!**
Game: [https://lichess.org/KAxNVVzc#56](https://lichess.org/KAxNVVzc#56)
Bad move: 29. Qa3, (Stockfish eval goes from -0.2 to #-8)
Correct move: f3
Configuration: lc0 cudnn - May 26 (default parameters), Windows 10 x64, Nvidia Titan V, Intel i5-7400T quad core, 32 GB RAM
Network ID: 346
Time control: 60" + 2"
Comments: **Leela was ahead but blundered and was checkmated**

**Comment by steve3140** [comment]
I just re-tested NN350 against MultiMove #1 position above.

lczero.exe -w weights.txt
position fen 6k1/4bpp1/2q1p2p/2p1P3/2P1N2P/2Bn1QP1/5P1K/8 b - - 0 35
go movetime 1000

info string   Qd7 ->      15   (V: 33.78%) (N: 39.14%) PV: Qd7 Nd6 Nxe5 Bxe5 Bxd6 Qd3 Qc6 Qxd6
info string   Nb4 ->      68   (V: 47.26%) (N: 31.72%) PV: Nb4 Kg2 Qa6 Nd2 Nc6 Qe4 Bf8
info string stm Black winrate 43.84%

That looks okay at that point, but NN350 still recommends the losing 35...Nb4 and doesn't notice the problems with it until 58kN.

She recommends the saving 35...Qd7 only after 129kN (NN316 found it in 19kN, NN311 in 29kN).

So tactically, at least in this case, things seem to have gone backwards again?  Bit disappointing.

**Comment by ASilver** [comment]
Ok, this was frokm CLOP test, and really bad. In position below, Bh3+ wins queen on the spot. Even on my quad with GTX1060, LCZero v0.10 (default settings) and id358 take 3m30s to see this!

![358-bh3](https://user-images.githubusercontent.com/3589847/40735866-c0190124-6412-11e8-902c-ac3054d7ce6b.jpg)

Here is full PGN. Key move is move 28.

[Event "?"]
[Site "?"]
[Date "2018.05.28"]
[Round "1"]
[White "lc0-may22"]
[Black "ice3"]
[Result "0-1"]
[ECO "A05"]
[PlyCount "65"]
[EventDate "2018.??.??"]
[TimeControl "60+1"]

1. Nf3 {book} Nf6 {book} 2. g3 {book} g6 {book} 3. Bg2 {book} Bg7 {book} 4. O-O
{book} O-O {book} 5. d3 {book} c5 {book} 6. c4 {0.02/2 3.3s} Nc6 {-0.09/19 4.6s
} 7. Nc3 {0.08/2 2.3s} d6 {0.00/99 0.015s} 8. Rb1 {0.08/2 4.1s} a6 {-0.20/18 2.
1s} 9. a3 {0.09/2 2.5s} Rb8 {-0.11/20 2.1s} 10. b4 {0.08/2 1.8s} cxb4 {-0.12/
20 2.1s} 11. axb4 {0.06/1 1.2s} b5 {-0.14/20 2.0s} 12. cxb5 {0.08/2 1.0s} axb5
{-0.14/18 2.0s} 13. Be3 {0.12/2 1.8s} Bd7 {-0.14/16 1.9s} 14. d4 {0.26/1 2.7s}
Rc8 {0.09/18 2.0s} 15. Qd2 {0.17/2 3.1s} Bf5 {-0.09/17 1.9s} 16. Rb3 {0.09/2 2.
0s} d5 {0.14/17 1.6s} 17. Rc1 {0.19/2 3.0s} Ne4 {0.02/20 1.9s} 18. Qd1 {
0.26/1 2.1s} Nxb4 {0.13/20 1.6s} 19. Nxe4 {0.26/2 1.6s} dxe4 {0.12/22 1.9s} 20.
Rxc8 {0.22/1 1.3s} Bxc8 {0.03/22 1.9s} 21. Rxb4 {0.33/2 2.0s} exf3 {-0.06/22 1.
6s} 22. Bxf3 {0.33/1 1.2s} Bd7 {-0.05/21 2.0s} 23. Kg2 {0.36/2 2.7s} Qb6 {
0.47/18 1.8s} 24. Qd3 {0.39/1 3.1s} Rc8 {0.43/19 1.9s} 25. Rb3 {0.38/2 3.0s} h5
{0.67/18 1.9s} 26. h4 {0.29/1 3.0s} Qd6 {0.43/18 1.7s} 27. Qe4 {0.38/1 1.9s}
Qc7 {0.04/19 3.1s} 28. Qxe7 {1.54/1 1.7s} Bh3+ {9.96/23 1.7s} 29. Kxh3 {
-10.24/2 1.9s} Qxe7 {9.92/23 1.7s} 30. Kg2 {-9.08/2 1.8s} b4 {10.57/22 1.7s}
31. Bd2 {-8.79/2 1.8s} Bxd4 {10.99/22 1.6s} 32. Rxb4 {-7.64/2 2.6s} Bxf2 {
11.62/24 2.9s} 33. Kxf2 {adjudication -9.21/2 1.9s, Black wins by adjudication}
0-1

**Comment by so-much-meta** [comment]
@ASilver... Very similar to one of the match game blunders I referenced above (although this was a little bit older)...
"""
White missed a simple tactic to capture black's queen on move 42, would have won but goes on to lose:
http://www.lczero.org/match_game/311127
"""

All of these types of issues seem to be due to Leela having very small priors when it *appears* that a strong piece is left under attack and undefended. In my opinion, simple 1 and 2 move tactics should set the bar for how flat the training target policies are. If training is currently unable to learn that it *might* be good to put the opponent's king in check in order to capture a queen, then IMO, that's a clear signal to flatten out the policies.... For example, making the PUCT/FPU changes you offered should help. I think, perhaps, people are too concerned about things that worked in Leela Zero. But chess has a much more jagged landscape than Go, full of terrain with very sharp ups and downs. It's very hard for any value head to smooth that out completely. IMO, the policy entropy needs to be increased just to give the value head a chance.

Anyway, I'm hoping that eventually the devs will agree that the train target policy entropy is too small, as well as the entropy fo the policy head (which I plan to do analysis on), and that corrections will be made to flatten this out and make it less sharp.

**Comment by tooweaktooslow** [comment]
![](https://cdn.discordapp.com/attachments/430695662108278784/453221264119693324/unknown.png)
ID367 doesn't see mate in two (I left her running for a bit, and she finally saw it at over 1 million nodes, which is a ""bit"" too many for a mate in two.)
Default lc0 settings.


![](https://i.imgur.com/O0Ea6sf.jpg)
It takes ID 367 over 30k nodes to see that giving a knight for no reason isn't good. (all the while giving herself +5 eval)
Default lc0 settings.


![](https://i.imgur.com/oWMW7xB.jpg)
3r4/1p3pkp/1qp5/6p1/Pp2Pp1P/1P3Pn1/2Q3PK/2BR4 b - - 0 29 
ID 367 doesn't see the Qf2 tactic, even with 8 million nodes.
Default lc0 settings.


![](https://i.imgur.com/b3737xN.jpg)
r3k2r/pR2pp1p/6p1/8/b2bP3/8/q2BBPPP/3Q1RK1 w kq - 2 16 
ID 367 doesn't see the Rxe7 tactic, even with 8 million nodes.
Default lc0 settings.

**Comment by mooskagh** [comment]
Please use forms from the first message of the issue.
Just posting screenshots with comments "it doesn't see anything" is not that useful.

Also, if you do some preparation work (like importing into lichess, pointing what's the correct move and what wrong move is done instead), you'll save time to (hopefully) multiple people who will look into that, and they won't have to spend time on that multiple times.

**Comment by tooweaktooslow** [comment]
i provided FENs for the tactical positions, so a lichess link isn't necessary
correct moves can be seen in the screenshot, which should be looked at as it's the whole point of the post
configuration and id is in the post

**Comment by dubslow** [comment]
Having the screenshot for extra details is fine, but for people trying to collate the entire contents of the thread, having one standard format makes that job much easier than trying to parse the unique format of your post, for N different "you" people making posts.

**Comment by Killua-** [comment]
I see, I didn't think of that. I'll stick to the format from now on, thanks.

**Comment by rwbc** [comment]
The unique standard is FEN and this is provided.
Nothing else is also needed for further processing in whatever way.

**Comment by tooweaktooslow** [comment]
![image](https://user-images.githubusercontent.com/37477460/40993976-41187466-68fb-11e8-9bd1-217754cbb14e.png)


Game: https://lichess.org/study/yyIYpmle (game 4)

Bad move: 81. b7??

Correct move: any bishop move that stays on the b1-h7 diagonal

Configuration: default win-lc0 on a gtx1070

Network ID: 367

Time control: 40/40, taken 68s for the move (> 500k nodes)

Comments: She realized that she blundered immediately (eval -40 next move), so she either overlooked Rxe4 completely (very unlikely given she used 500k nodes), or she thought she could promote one of the pawns after b7. However, the rook does a very good job of threatening mate and indirectly covering the promotions (if h8=Q, Ra4#; if b8=Q, Ra4+ followed by Rb4+ taking the queen. That's altogether a 3 move tactic which she definitely should've seen in 500k nodes. 
I tested her with PUCT=3.0 and she still didn't see Rxe4.
![image](https://user-images.githubusercontent.com/37477460/40994463-8a6557b4-68fc-11e8-97ad-4b1f242340a9.png)
